### PR TITLE
feat: `onUpdate` hook for apps

### PR DESCRIPTION
**Pull request by KevLehman**
# What? :boat:

Added hook to be called when an app is updated

# Why? :thinking:

Because why not 🤷🏽‍♂️ 
Serious: allows apps to do stuff only when an update happens, like sending messages or create cron jobs.
# Links :earth_americas:

Related PR to get `user` on context: https://github.com/RocketChat/Rocket.Chat/pull/32719
# PS :eyes:

### CODE REVIEW

File: src/server/AppManager.ts [hunk 1]
```diff
@@ -718,6 +718,8 @@ export class AppManager {
                 .catch(() => {});
         }
 
+        await this.updateApp(app)
```
Comment (KevLehman) [inline src/server/AppManager.ts:721]:
```suggestion
        await this.updateApp(app);
```

**Comment by d-gubert** [comment]
Thanks for this! 

I just think we're missing here the signature for the method in the `App` class, like the [`onInstall` event here](https://github.com/RocketChat/Rocket.Chat.Apps-engine/blob/3a59d4018a94026bd269bc569fa31f28a92a5a63/src/definition/App.ts#L182) for example. Regarding contextual information about the event, I'd say it would be cool to know the user who triggered the update and maybe what the previous version was? I'm open to ideas :)

File: src/definition/accessors/IAppUpdateContext.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import type { IUser } from '../users';
+
+export interface IAppUpdateContext {
+    user: IUser | null;
```
Comment (d-gubert) [inline src/definition/accessors/IAppUpdateContext.ts:4]: This looks like an optional property, is there a reason why you didn't use `user?: IUser`?

Comment (KevLehman) [reply src/definition/accessors/IAppUpdateContext.ts:4]: Hmmm, good question. Cannot remember 🤔 let me double check

Comment (d-gubert) [reply src/definition/accessors/IAppUpdateContext.ts:4]:
```suggestion
    user?: IUser;
```

**Comment by d-gubert** [comment]
Example usage:

```
    public async onUpdate(context: IAppUpdateContext, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify): Promise<void> {
        console.log("Hello from update", context);

        const message = modify.getCreator().startMessage();

        const room = await read.getRoomReader().getById("GENERAL");

        if (!room) {
            throw new Error('oh no');
        }

        message.setText("Hello, thank you for updating!").setRoom(room);

        await modify.getCreator().finish(message);
    }

File: src/server/AppManager.ts [hunk 3]
```diff
@@ -650,7 +650,11 @@ export class AppManager {
         this.apps.delete(app.getID());
     }
 
-    public async update(appPackage: Buffer, permissionsGranted: Array<IPermission>, updateOptions = { loadApp: true }): Promise<AppFabricationFulfillment> {
+    public async update(
+        appPackage: Buffer,
+        permissionsGranted: Array<IPermission>,
+        updateOptions: { loadApp?: boolean; user: IUser | null } = { loadApp: true, user: null },
```
Comment (d-gubert) [inline src/server/AppManager.ts:656]:
```suggestion
        updateOptions: { loadApp?: boolean; user?: IUser } = { loadApp: true },
```

**Review by d-gubert — APPROVED** [review]

**Pull request merged**